### ISSUE TITLE
Question: Why does generating a probability not give a definite number?

### ISSUE THREAD

**Issue by ngzhiyang** [BrainJS/brain.js#516]
![image](https://user-images.githubusercontent.com/20526190/75956664-51c2b480-5ef3-11ea-910c-03283fc9acbe.png)

## *What* is wrong?
Why does generating a probability not give a definite number? I don't understand the mechanisms behind Brain.js and I'm new to machine learning but I'm interested to know is this only for Brainjs? or for every neural network? doesn't a fix data set produce a definite probability of output ?

## *How* important is this (1-5)?

2

## Other Comments
Thank you for your time and attention

**Comment by robertleeplummerjr** [comment]
You'll have to be more descriptive.  Can you include a code sample?

A recurrent network does give a definite probability because it's dialect is exactly what it was given, but feedforwars and timestep will not because as you state, they are probabilities.

Have you ever heard a weather man say: "It will be exactly 72.5555555556 degrees tomorrow"?  No.  Because he is estimating using similar (likely much more sophisticated) software.  Nobody knows the future, but they can guess pretty good.

To further the point, take a camera and take a picture, now take a picture of that picture, can you tell the difference between the original and the picture?  It is a very good copy, which is essentially what the network is doing.  It is measuring error against what it was input.  Internally it will really never know what it was actually given, it only measures how close it was.

With 32 bit floating point accuracy, there is nearly an infinite gap between 0 and 1 to the computer.

To even further the point.  Make a sculpture of a lion out of clay, try your hardest to make it perfect.  Is it a lion?  Sure.  Is it an exact copy?  No.  Now make the sculpture of the sculpture, each time it degrades in quality.  Because you are measuring error and probability on a practical (depending or not if your spouse calls you out on it) level.  Each time the sculpture looses something, precision.

**Comment by ngzhiyang** [comment]
Hi Robert, 

Thank you for your prompt reply. I'm confused, won't a recurrent network give a definite output/result not a definite probability? 

Here is the code example:
const network = new brain.NeuralNetwork();
network.train([
    {input: [1,2], output: [1]}, // Team 2 wins
    {input: [1,3], output: [1]}, // Team 3 wins
    {input: [2,3], output: [0]}, // Team 2 wins
    {input: [2,4], output: [1]}, // Team 4 wins
    {input: [1,2], output: [0]}, // Team 1 wins
    {input: [1,3], output: [0]}, // Team 1 wins
    {input: [2,4], output: [0]}, // Team 2 wins
    {input: [1,5], output: [0]}, // Team 1 wins
]);
const output = network.run([1,4]);
console.log(`Prob : ${Output}`);

If there are a fixed number of cases, then the probability of team 4 winning against team 1 should remain constant?

I have some other minor questions I'm hoping you could provide some insight into and possibly clear up my misunderstanding:
1. How come if I run [1,5] that is in the training set of data, the probability is not = 1 , isn't the only case of only  [1,5] given? My guess is that because 'Team 1' has been implicated in other I/Os .. then what if its a separate case with different variables?
2. Does running the output multiple times make the process of retrieving faster? E.g. if I run 'const output = network.run([1,4])', many times, will it make its process of retrieval of the result faster?
3. In a recurrent neural network, why does more iterations make it more accurate when the process is the same when theres no new I/O to the training data, is it because of its property of being a LSTM?

**Comment by robertleeplummerjr** [comment]
>I'm confused, won't a recurrent network give a definite output/result not a definite probability?

The output is _heavily_ related to "probability".

>Here is the code example:

What does your training data mean?  I'm not sure I interpret it correctly.

>1. How come if I run [1,5] that is in the training set of data, the probability is not = 1

I think the net doesn't understand your training data.

>Does running the output multiple times make the process of retrieving faster?

No.

>In a recurrent neural network, why does more iterations make it more accurate

A feed forward net simply learns differently than a feed forward.  As well the output error is (in the case of a LSTM network) measured differently.
